### PR TITLE
Scopes and Password Work

### PR DESCRIPTION
**Pull request by efatsi**
Article scopes in place, can view articles and comments posted by a specific user, password reset, lost password, and 'remember me' button available, date picker for article view also working

### CODE REVIEW

File: app/controllers/articles_controller.rb [hunk 1]
```diff
@@ -1,13 +1,39 @@
...
   
   before_filter :assign_article, :only => [:show, :edit, :update, :destroy]
-  before_filter :require_user, :except => [:index, :show]
+  before_filter :require_user, :only => [:new, :edit]
 
   load_and_authorize_resource
 
   def index
-  	@articles = Article.chronological.search(params[:search], params[:page])
+
```
Comment (brianjlandau) [inline app/controllers/articles_controller.rb:9]: All this parameter parsing should be broken out into a search class it shouldn't be in the controller.

File: app/controllers/articles_controller.rb [hunk 2]
```diff
@@ -48,6 +74,28 @@ def create
       end
     end
   end
+  
+  def pick_date
```
Comment (brianjlandau) [inline app/controllers/articles_controller.rb:78]: What does this do?

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -6,18 +6,23 @@ def new
   end
   
   def create
-    if User.authenticate(params[:username], params[:password])
-    	session[:user_id] = User.find_by_username(params[:username]).id
-    	flash[:notice] = "Logged In!"
+    user = User.authenticate(params[:username], params[:password])
+    if user
+      if params[:remember_me]
+        cookies.permanent[:auth_token] = user.auth_token
```
Comment (brianjlandau) [inline app/controllers/sessions_controller.rb:12]: HOT I didn't know about this.

File: app/models/article.rb [hunk 4]
```diff
@@ -1,12 +1,14 @@
 class Article < ActiveRecord::Base
   
   URI_REGEX = /\A(http|https):\/\/([a-z0-9]*[\-\.])?([a-z0-9]+*\.[a-z]{2,5})(:[0-9]{1,5})?(\/.*)?\z/
-  
+  YEARS = (Time.now.year-5..Time.now.year).map{|y| y}.reverse
+  MONTHS = Hash[["January", "Fubruary", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"].zip [1,2,3,4,5,6,7,8,9,10,11,12]]
+  DAYS = (1..31).map{|a| a}
```
Comment (brianjlandau) [inline app/models/article.rb:6]: `(1..31).to_a`

File: app/models/article.rb [hunk 5]
```diff
@@ -38,6 +40,14 @@ def short_link
   end
   
   # Scopes
-  scope :chronological, :order => 'created_at DESC'
+  scope :chronological, order('created_at DESC')
+  scope :by_user, lambda { |user_id| where("user_id = ?", user_id) }
+  scope :by_rating, order('points DESC, created_at DESC')
+  scope :this_day, where('created_at > ?', Time.now - 1*60*60*24)
```
Comment (brianjlandau) [inline app/models/article.rb:46]: Use time number helpers instead of these (i.e. 1.day.ago)

Comment (brianjlandau) [inline app/models/article.rb:46]: Actually shouldn't these be in lambda's so the time is calculated when the scope is run not when the class is loaded.

File: app/models/user.rb [hunk 6]
```diff
@@ -14,6 +16,19 @@ class User < ActiveRecord::Base
...
 	# Methods
+
+  def send_password_reset
+    generate_token(:password_reset_token)
+    self.password_reset_sent_at = Time.zone.now
+    save!
+    UserMailer.password_reset(self).deliver
+  end
+
+  def generate_token(column)
```
Comment (brianjlandau) [inline app/models/user.rb:27]: :+1:

File: config/environments/development.rb [hunk 7]
```diff
@@ -31,7 +31,16 @@
 
   # Do not compress assets
   config.assets.compress = false
-
+  
   # Expands the lines which load the assets
-  config.assets.debug = true
+  # This needs to be false for bootstrap to work in development
```
Comment (brianjlandau) [inline config/environments/development.rb:36]: That's seriously weird. that shouldn't be true, I don't recall having to do that with the bootstrap gem I've used.

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -34,12 +34,31 @@ def create
...
     else
       render :action => "edit"
     end
   end
   
+  def change_password 
+    @user = User.find(params[:id]) 
+    if request.post? 
+      if User.authenticate(@user.username, params[:password][:old_password]) == @user 
+        @user.password = params[:password][:new_password]
```
Comment (brianjlandau) [inline app/controllers/users_controller.rb:47]: Some of this should probably be put into a method on `User`.

File: config/environments/development.rb (hunk 7 above)
Comment (efatsi) [inline config/environments/development.rb:36]: Ya I don't know why. The same thing happened with the card game app I'm building now though, maybe I'm not setting up bootstrap correctly?

**Pull request merged**